fl4health/strategies/flash.py

```diff
@@ -0,0 +1,175 @@
+                Client-side learning rate. Defaults to 1e-1.
+            beta_1 : float, optional
+                Momentum parameter. Defaults to 0.9.
+            beta_2 : float, optional
+                Second moment parameter. Defaults to 0.99.
+            tau : float, optional
+                Controls the algorithm's degree of adaptability. Defaults to 1e-9.
+            weighted_aggregation (bool, optional): Determines whether parameter aggregation is a linearly weighted
+                average or a uniform average. Flash default is a uniform average by the number of clients.
+                Defaults to False.
+            weighted_eval_losses (bool, optional): Determines whether losses during evaluation are linearly weighted
+                averages or a uniform average. Flash default is a uniform average of the losses by dividing
+                the total loss by the number of clients. Defaults to False.
+        """
+
+        super().__init__(
+            fraction_fit=fraction_fit,
+            fraction_evaluate=fraction_evaluate,
+            min_fit_clients=min_fit_clients,
+            min_evaluate_clients=min_evaluate_clients,
+            min_available_clients=min_available_clients,
+            evaluate_fn=evaluate_fn,
+            on_fit_config_fn=on_fit_config_fn,
+            on_evaluate_config_fn=on_evaluate_config_fn,
+            accept_failures=accept_failures,
+            initial_parameters=initial_parameters,
+            fit_metrics_aggregation_fn=fit_metrics_aggregation_fn,
+            evaluate_metrics_aggregation_fn=evaluate_metrics_aggregation_fn,
+            weighted_aggregation=weighted_aggregation,
+            weighted_eval_losses=weighted_eval_losses,
+        )
+        self.current_weights = parameters_to_ndarrays(initial_parameters)
+        self.eta = eta
+        self.eta_l = eta_l
+        self.tau = tau
+        self.beta_1 = beta_1
+        self.beta_2 = beta_2
+        self.m_t: Optional[NDArrays] = None
+        self.v_t: Optional[NDArrays] = None
+        self.d_t: Optional[NDArrays] = None
+
+    def __repr__(self) -> str:
+        """Compute a string representation of the strategy."""
+        rep = f"Flash(accept_failures={self.accept_failures})"
+        return rep
+
+    def _update_parameters(self, delta_t: NDArrays) -> None:
+        """Update m_t, v_t, beta_3, and d_t per-element"""
+        assert self.v_t is not None and self.m_t is not None and self.d_t is not None
+
+        for i, (delta, m, v, d_prev) in enumerate(zip(delta_t, self.m_t, self.v_t, self.d_t)):
+            # Update m_t
+            self.m_t[i] = self.beta_1 * m + (1 - self.beta_1) * delta
+
+            # Update v_t
+            self.v_t[i] = self.beta_2 * v + (1 - self.beta_2) * np.square(delta)
+
+            # Compute beta_3
+            norm_v_prev = np.abs(v)
+            norm_diff = np.abs(np.square(delta) - self.v_t[i])
+            beta_3_matrix = norm_v_prev / (norm_diff + norm_v_prev)
```

Comment:
Should these be euclidean norms as per line 19 in Algorithm 1 in the paper? I could be missing something though I see you and David went back and forth about similar part of the code earlier

Reply:
In line 19 of the algorithm, the best we can tell is that the operations are being carried out element-wise (using the index j). The l2-norm of a single element is just the absolute value ($\Vert a \Vert_2 = \sqrt{a^2} = \vert a \vert$).

Reply:
Ahh, I assumed j was an index over layers or something but I see from the discussion from before we have reason to believe it is not. Seems pretty weird to have the l2-norm in the algorithm since its elementwise though haha

Reply:
Yeah I agree. There isn't any indication that $\Delta^{(5)}$ or any of the other parameters are structured element-wise that I could find. If they didn't have a broken link to their code, we could verify, but it's not there 🤦

Reply:
unfortunately the paper was super unclear when it came to the updating, and its never explicitly stated what `n` stands for, the only logical conclusion was is the update happpens per-element, which meant the norms of each element is equivalent to the abs for the matrix